### PR TITLE
Frontend rework

### PR DESCRIPTION
**Pull request by aurarius1**
As said in #295 here the draft request for the frontend updates.

I did include a revised version of the Dockerfile and tested building an image with it for the Pi Zero (then copied it on the Pi Zero, since building on it directly will not work anymore due to the issues with the updated frontend I already outlined in the discussions). The resulting image is less than half in size. Nevertheless since raspberrypi.com suggests using a 16GB SD Card for Raspberry Pi OS Lite I updated the requirements a little bit, to ensure it works smoothly for everybody (since technically we cannot do much if the remaining storage is too low for the image to run). 

I also included the PSA that docker dropped support for the Pi Zero. 

This is opened as a draft, because it might still need some work and I think it is easier to checkout a pull request than fork a different repo.

### CODE REVIEW

**Comment by DarthArnettJr** [comment]
Okay, now. Let’s hope that this method won’t be heck to set up now, eh?

**Comment by aurarius1** [comment]
> Okay, now. Let’s hope that this method won’t be heck to set up now, eh?

I don’t think that’s a fair assessment. Most of the issues you’ve run into stem from the very constrained storage environment. I’ve set this up on several Raspberry Pi models during testing, and it worked fine in each case.

**Comment by DarthArnettJr** [comment]
Well I hope I can get my hands on a 32 GB micro SD card from Target, their same day pickup is a GODSEND.

**Comment by aurarius1** [comment]
Don't know if anybody else had time to test the branch as well. Should i change base of the pull request such that it will be merged into dev first?

File: docs/pages/getting-started/troubleshooting.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+After the installation has finished, exit Zadig and restart RPCS3. If you get stuck on the main menu, just close the game, right-click on it in the RPCS3 games list, select `Change Custom Configuration`, switch to the `Network` tab and choose `Disconnected` in both drop-down menus. The game will now correctly detect the Toy Pad.
+
+To undo the changes from Zadig, you have to rollback the driver:
+
+1. Open `Device Manager`, scroll down to `USB devices` and expand the section.
+2. Double-click `LEGO READER V2.10`.
+3. Switch to the `Driver` tab, click `Previous Driver`, select the first option and click yes.
+
+**Solution for Linux systems**
```
Comment (Berny23) [inline docs/pages/getting-started/troubleshooting.md]: Should be a header, similar to above:
`### Linux solution`

**Review by Berny23 — CHANGES_REQUESTED** [review]
Great work! Some small things I noticed.

**Review by Berny23 — APPROVED** [review]

**Comment by DarthArnettJr** [comment]
A thing I'd like to say is that the ToyPad disconnects because it's USB pulse is off-set on a Zero W, The fix for me was to set it to tmpfs.

**Comment by Berny23** [comment]
Yeah, I think the game is generally pretty bugged, even with original hardware. It also crashed frequently in RPCS3 back when I completed it 100 % with the VM method.

Older Lego games have a tendency to crash and glitch, it happened to me with Lego Marvel Super Heroes on my real Wii U. It wouldn't surprise me, if Dimensions had some kind of sync issue or race condition with portal communication.

**Comment by Dave247** [comment]
I noticed you added some simple temporary instructions @aurarius1 back in Jan on how to how to run your build which I previously had no idea on how to do. But unfortunately I still don't seem to be smart enough to be able to figure out which line I need to modify/add for the fourth step.

I think it is line 18, but I am not sure what the correct syntax is supposed to be?

**Comment by aurarius1** [comment]
> I noticed you added some simple temporary instructions @aurarius1 back in Jan on how to how to run your build which I previously had no idea on how to do. But unfortunately I still don't seem to be smart enough to be able to figure out which line I need to modify/add for the fourth step.
> 
> I think it is line 18, but I am not sure what the correct syntax is supposed to be?

I used these commands on my desktop and then exported the built image and copied it to the Pi using SSH: 

```bash
sudo podman run --rm --privileged docker.io/multiarch/qemu-user-static --reset -p yes 
podman build \                                   
  --platform=arm64 \      
  -t ld-toypad-emulator:latest .
```

you just need to adapt the platform tag. The first lien is to allow podman to build for different platforms. 

Then you can export it like that: 

```bash
 podman save -o ld-toypad-emulator-arm64.tar localhost/ld-toypad-emulator:latest 
 ```
 
 and copy it via ssh to your pi and import it with this command: 
 
 ```bash
 podman load -i ld-toypad-emulator-arm64.tar
 ```

**Comment by aurarius1** [comment]
I just wanted to give a quick update: I think I found the issue with the "vanishing" frontend. It is most likely also related to the asynchronous file writing in the backend. I will add proper redirections to error pages, such that the frontend won't disappear due to this. I think the backend should not be touched by this pull request, but those issues should be fixed afterwards as well. 

I'll test a bit with the error page, etc., before pushing the code!

**Comment by Berny23** [comment]
Please note, there has been lots of code cleanup in #326.
You may have to rebase or merge.

**Comment by aurarius1** [comment]
Thanks for the heads up. Sorry that it is taking so long to push the error handling stuff, I really couldn't test it yet, life is busy atm.

**Comment by Dave247** [comment]
I got distracted/busy with other things as well and haven't touched anything for a while. Decided to buy a Zero 2 to use exclusive for this project as I wanted to use my 4 for Steam Link, which was difficult given how many times I have been re-flashing the OS.

Trixie and Bookworm are still busted (as noted in #325) but that's no surprise, given that some time has passed I wanted to refresh myself on where things were.

I've spent a lot of hours recently trying (and failing) to get legacy GPU drivers working in Debian 13, which got me to think that if it's the USB gadget side that is failing then it's probably because whatever drivers and libraries the project was originally used aren't supported in newer kernel versions.
Like for example how RPI OS Trixie now has its own built in [rpi-usb-gadget](https://github.com/raspberrypi/rpi-usb-gadget) package that maybe the project should switch to using instead.

Given that this was meant to be for your purely frontend work @aurarius1 and it's kinda my fault for opening a can of can of worms and getting thigs off track. I agree it's best for you to just focus on just the frontend stuff which is what you already started, as the rest really needs to get collated and split off into its own thing.

**Comment by aurarius1** [comment]
I just merged everything back into the frontend branch and will test everything next week. After that it should be done! Sorry for taking so long, a lot is going on rn.

**Comment by aurarius1** [comment]
From what I've seen so far, everything still works except the issues with creating tokens. Is it okay if i change the methods to edit the toytags.json synchronously? 

Also, what I just noticed, if the game writes update commands to a vehicle tag (tested when creating the Batmobile for the first time), the tags will have N/A stored as the name in the toytags.json, which also happens with the original frontend. However, the original frontend doesn't rely on the name stored in toytags.json, mine does. I found that the `getAnyNameFromId` function seems to have the ID check backwards, if looks up the name from the vehicle map if the ID is `< 1000`, else from the character map. From the IDs stored in the two maps, I think it should be the other way round, right? If so, is it also okay if I make this change with this pull request? 

EDIT: just saw with the original frontend, names with N/A are filtered out on a page refresh. This might also be the reason for #280

**Comment by aurarius1** [comment]
Instead of further relying on toytags.json, the project could also implement a sqlite database for created toytags. However, if this is the way to go, refactoring this should definitely be a new PR.

**Comment by Berny23** [comment]
Thanks for looking into it. You can include the json fixes here too.

**Comment by aurarius1** [comment]
I've played the first level with the changes, seems to work fine for me. However, somebody else should test this as well. 

I did not add the batch creation route, but the file reads are now synchronous.
